File: services/client/src/protocol/protocol.go
package protocol

import (
	"io"
	"strings"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const HEADER_SIZE = 2
const BATCH_SEPARATOR = "\n"
const BATCH_DEFAULT_SIZE = 16

var ACK_MARKER = []byte("OK")

func putUint16BigEndian(b []byte, v uint16) {
	b[0] = byte(v >> 8)
	b[1] = byte(v)
}

func uint16BigEndian(b []byte) uint16 {
	return uint16(b[0])<<8 | uint16(b[1])
}

func EncodeFrame(payload []byte) []byte {
	frame := make([]byte, HEADER_SIZE, HEADER_SIZE+len(payload))
	putUint16BigEndian(frame, uint16(len(payload)))
	return append(frame, payload...)
}

func ReadFrame(r io.Reader) ([]byte, error) {
	header, err := safe_socket.RecvAll(r, HEADER_SIZE)
	if err == io.EOF {
		return nil, io.EOF
	}
	if err != nil {
		return nil, err
	}
	size := int(uint16BigEndian(header))
	payload, err := safe_socket.RecvAll(r, size)
	if err != nil {
		return nil, err
	}
	return payload, nil
}

func MarshalBet(agency string, fields []string) []byte {
	record := append([]string{agency}, fields...)
	return []byte(strings.Join(record, ","))
}

func MarshalBatch(agency string, records [][]string) []byte {
	lines := make([]string, 0, len(records))
	for _, record := range records {
		lines = append(lines, string(MarshalBet(agency, record)))
	}
	return []byte(strings.Join(lines, BATCH_SEPARATOR))
}

func UnmarshalBet(payload []byte) []string {
	return strings.Split(string(payload), ",")
}

func UnmarshalBatch(payload []byte) [][]string {
	lines := strings.Split(string(payload), BATCH_SEPARATOR)
	records := make([][]string, 0, len(lines))
	for _, line := range lines {
		records = append(records, UnmarshalBet([]byte(line)))
	}
	return records
}

File: services/server/src/protocol/protocol.py
import socket

import safe_socket

HEADER_SIZE = 2
BATCH_SEPARATOR = "\n"

ACK = b"OK"

def encode_frame(payload: bytes) -> bytes:
    length = len(payload).to_bytes(HEADER_SIZE, byteorder="big")
    return length + payload


def read_frame(sock: socket.socket):
    header = safe_socket.recv_all(sock, HEADER_SIZE)
    if not header:
        return None
    if len(header) < HEADER_SIZE:
        raise ConnectionError("truncated message header")
    payload_size = int.from_bytes(header, "big")
    payload = safe_socket.recv_all(sock, payload_size)
    if len(payload) < payload_size:
        raise ConnectionError("truncated message payload")
    return payload

def marshal_bet(agency_id, fields) -> bytes:
    return ",".join([str(agency_id), *fields]).encode()


def unmarshal_bet(payload: bytes):
    return payload.decode().split(",")


def unmarshal_batch(payload: bytes):
    lines = payload.decode().split(BATCH_SEPARATOR)
    return [unmarshal_bet(line.encode()) for line in lines]